### ISSUE TITLE
Order Adjustment not found

### ISSUE THREAD

**Issue by wilmer05** [craftcms/commerce#2095]
1. After the checkout with a klarna plugin the plugin tries to save a last transaction and fails

**Additional info**
- Craft version: 3.5.15.1
- Commerce version: 3.2.11
- PHP version: 7.1

The stacktrace I get is the following

> 2021-04-20 16:24:38 [-][-][q7i0eucabhet0rv8lhulklrnad][error][yii\base\Exception] yii\base\Exception: Inge
n ordrejustering eksisterer med ID-en «845845» in /var/www/123apotek/src/vendor/craftcms/commerce/src/serv
ices/OrderAdjustments.php:178
Stack trace:
#0 /var/www/123apotek/src/vendor/craftcms/commerce/src/elements/Order.php(3105): craft\commerce\services\O
rderAdjustments->saveOrderAdjustment()
#1 /var/www/123apotek/src/vendor/craftcms/commerce/src/elements/Order.php(1963): craft\commerce\elements\O
rder->_saveAdjustments()
#2 /var/www/123apotek/src/vendor/craftcms/cms/src/services/Elements.php(2461): craft\commerce\elements\Ord
er->afterSave()
#3 /var/www/123apotek/src/vendor/craftcms/cms/src/services/Elements.php(764): craft\services\Elements->_sa
veElementInternal()
#4 /var/www/123apotek/src/vendor/craftcms/commerce/src/elements/Order.php(1458): craft\services\Elements->
saveElement()
#5 /var/www/123apotek/src/vendor/craftcms/commerce/src/services/Transactions.php(429): craft\commerce\elem
ents\Order->updateOrderPaidInformation()
#6 /var/www/123apotek/src/vendor/elleracompany/commerce-klarna-checkout/src/controllers/KlarnaController.p
hp(97): craft\commerce\services\Transactions->saveTransaction()
#7 [internal function]: ellera\commerce\klarna\controllers\KlarnaController->actionConfirmation()
#8 /var/www/123apotek/src/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#9 /var/www/123apotek/src/vendor/yiisoft/yii2/base/Controller.php(180): yii\base\InlineAction->runWithPara
ms()
#10 /var/www/123apotek/src/vendor/craftcms/cms/src/web/Controller.php(190): yii\base\Controller->runAction
()
#11 /var/www/123apotek/src/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction()
#12 /var/www/123apotek/src/vendor/craftcms/cms/src/web/Application.php(274): yii\base\Module->runAction()
#13 /var/www/123apotek/src/vendor/craftcms/cms/src/web/Application.php(577): craft\web\Application->runAct
ion()
#14 /var/www/123apotek/src/vendor/craftcms/cms/src/web/Application.php(253): craft\web\Application->_proce
ssActionRequest()
#15 /var/www/123apotek/src/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleReq
uest()
#16 /var/www/123apotek/src/web/index.php(24): yii\base\Application->run()

**Comment by lukeholder** [comment]
Probably best to try and reproduce with the built in dummy or manual gateway, and it it only happens with klarna submit an issue to https://github.com/elleracompany/commerce-klarna-checkout/issues

If it happens in the dummy gateway I will re-open this issue.

**Comment by wilmer05** [comment]
We were investigating this issue and we found that it was a race condition. We did the Klarna + Webshipper integration that one can see on the klarna website and we had a webhook calling an endpoint in craft once the order reached WS through Klarna.

However the webhook was triggering an update on the order in craft and sometimes it was failing since the order was still a cart. The cart is converted into an order by the klarna plugging once it reaches the confirmation page and confirms that it was paid.

In other words, since the WS webhook reached craft before the user went to the confirmation page, an update to an order was happening when this order was still a cart, meanwhile on the klarna confirmation page, another update was happening to save the successful authorize transaction.
